oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizedClientManager.java

```diff
@@ -0,0 +1,81 @@
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Implementations of this interface are responsible for the overall management
+ * of {@link OAuth2AuthorizedClient Authorized Client(s)}.
+ *
+ * <p>
+ * The primary responsibilities include:
+ * <ol>
+ *  <li>Authorizing (or re-authorizing) an OAuth 2.0 Client
+ *  	by leveraging an {@link OAuth2AuthorizedClientProvider}(s).</li>
+ *  <li>Managing the persistence of an {@link OAuth2AuthorizedClient} between requests,
+ *  	typically using an {@link OAuth2AuthorizedClientRepository}.</li>
+ * </ol>
+ *
+ * @author Joe Grandja
+ * @since 5.2
+ * @see OAuth2AuthorizedClient
+ * @see OAuth2AuthorizedClientProvider
+ * @see OAuth2AuthorizedClientRepository
+ */
+public interface OAuth2AuthorizedClientManager {
+
+	/**
+	 * Attempt to authorize or re-authorize (if required) the {@link ClientRegistration client}
+	 * identified by the provided {@code clientRegistrationId}.
+	 * Implementations must return {@code null} if authorization is not supported for the specified client,
+	 * e.g. the associated {@link OAuth2AuthorizedClientProvider}(s) does not support
+	 * the {@link ClientRegistration#getAuthorizationGrantType() authorization grant} type configured for the client.
+	 *
+	 * @param clientRegistrationId the identifier for the client's registration
+	 * @param principal the {@code Principal} {@link Authentication} (to be) associated to the authorized client
+	 * @param request the {@code HttpServletRequest}
+	 * @param response the {@code HttpServletResponse}
+	 * @return the {@link OAuth2AuthorizedClient} or {@code null} if authorization is not supported for the specified client
+	 */
+	@Nullable
+	OAuth2AuthorizedClient authorize(String clientRegistrationId, Authentication principal,
+										HttpServletRequest request, HttpServletResponse response);
+
+	/**
+	 * Attempt to re-authorize (if required) the provided {@link OAuth2AuthorizedClient authorized client}.
+	 * Implementations must return the provided {@code authorizedClient} if re-authorization is not supported
+	 * for the {@link OAuth2AuthorizedClient#getClientRegistration() client} OR is not required,
+	 * e.g. a {@link OAuth2AuthorizedClient#getRefreshToken() refresh token} is not available OR
+	 * the {@link OAuth2AuthorizedClient#getAccessToken() access token} is not expired.
+	 *
+	 * @param authorizedClient the authorized client
+	 * @param principal the {@code Principal} {@link Authentication} associated to the authorized client
+	 * @param request the {@code HttpServletRequest}
+	 * @param response the {@code HttpServletResponse}
+	 * @return the re-authorized {@link OAuth2AuthorizedClient} or the provided {@code authorizedClient} if not re-authorized
+	 */
+	OAuth2AuthorizedClient reauthorize(OAuth2AuthorizedClient authorizedClient, Authentication principal,
```

Comment:
Is there a reason we need to support reauthorize separately? It seems that the argument resolver does not need to use reauthorize. I'm guessing that ServletOAuth2AuthorizedClientExchangeFilterFunction could be restructured so that it does not require reauthorize as well